docs/proposals/SIL-Utilities.md

```diff
@@ -0,0 +1,106 @@
+# Utilities for SIL optimizer passes
+This document lists a set of SIL utilities to be used by the Swift implementation of the SIL optimizer.
+
+The purpose of this proposal is to do a better planning of what utilities we need in SIL optimizer passes.
+
+We want to avoid a situation like we have in the C++ SIL optimizer, where a huge amount of (partly) redundant utilities exist. Many of those utilities have overlapping functionality,  are difficult to discover and are poorly designed. We want to do better in the Swift SIL optimizer.
+
+## Basic Data-Structures
+#### `Stack`
+
+An allocation-free, array like data structure. To be used instead of `Swift.Array` wherever random-access is not needed.
+
+**Related C++ utilities:** `llvm::SmallVector`, `Stack`
+**Status:** done  
+
+#### `BasicBlockSet`
+An extremely efficient implementation of a set of basic blocks.
+
+**Related C++ utilities:** `BasicBlockSet`
+**Status:** done  
+
+#### `BasicBlockWorklist`
+To be used for all kind of basic-block work-list algorithms.
+
+**Uses:** `Stack`, `BasicBlockSet`  
+**Related C++ utilities:** `BasicBlockWorklist`
+**Status:** done  
+
+## Building SIL
+
+#### `Builder.insert(after:)`
+Useful to insert instructions after a (potential) terminator instruction.
+
+**Related C++ utilities:** `SILBuilderWithScope::insertAfter()`
+**Status:** done  
+
+## Instruction classifications
+We want to classify certain instructions into common groups so that passes can deal deal with the group instead of individual instruction opcodes. Optimization passes then don't need to be updated if new instructions are added to a group.
+
+In Swift we can easily do this by introducing protocols to which instruction classes can conform to.
+
+#### `ApplySite`
+
+**Related C++ utilities:** `ApplySite`
+**Status:** exists; to-do: complete member variables/functions  
+
+#### `Cast` or `ForwardingCast`
```

Comment:
Maybe `ValueCast` or `CastInst`. An ownership API can determine whether it forwards ownership or implicitly borrows or escapes a pointer or whatever.

This does not apply to address types. And does not apply to aggregation: struct_extract or struct. We may also want `AggregateInst` and `ExtractInst` because it's those are so commonly handled.

We may want all classifications to have a naming convention. Maybe they should all end in `Inst` if they are indeed instruction wrappers.

The non-inst wrappers will be `FunctionArgument`, `Phi`, and `TerminatorResult`. Exhaustively:
```
switch silArgument {
  case FunctionArgument: ;
  case Phi ;
  case TerminatorResult 
}
```

We *might* also have a `BorrowScope` utility that either wraps an argument or a begin_borrow

Reply:
Maybe call those protocol with an `Instruction` at the end? This is consistent with `SingleValueInstruction` (which is also not a concrete instruction type) and you can distinguish the name from a concrete instruction class.

Reply:
For arguments we already have `FunctionArgument` and `BlockArgument`, which has `isPhiArgument`.

Reply:
I pushed a new version

Reply:
The problem is, as long as we don't have separate phi- and non-phi block argument classes in C++ we cannot have distinct classes on the Swift side.

Also, I don't agree that it doesn't make sense, because a non-phi argument can become a phi-argument due to transformations in SimplifyCFG. This doesn't change anything about the argument itself, just its context (i.e. the predecessor terminator instruction) changes. So IMO, it makes sense to have a `isPhiArgument` property, which dynamically checks the situation.
Of course, SimplifyCFG could be careful and replace block arguments, but, hm, doesn't sound like it's worth the effort.

Reply:
It is essential that we have a Phi type and a TerminatorResult types. It is not essential that they be different opcodes. That only improves efficiency. If they are not separate opcodes, then SILFunctionArgument also should not be a separate opcode.

There is never a situation in which a pass cares where a SILArgument comes from that it doesn't need to distinguish phis from terminator results. They are handled in an opposite fashion in all the code I've seen. Having static types radically simplifies code that handles block arguments.

Wherever SimplifyCFG generates a new phi then it *definitely* should be explicitly creating a phi! I don't see any relationship between an old terminator result and a new phi.

SILBlockArgument is one of the most obvious reasons that we have so many confusing and redundant abstractions on the C++ side. The stated goal of this document is to correct that situation.

Reply:
I don't understand your comment. In Swift we don't have opcodes, we just have types.

> There is never a situation in which a pass cares where a SILArgument comes from that it doesn't need to distinguish phis from terminator results. They are handled in an opposite fashion in all the code I've seen.

I clearly misunderstood something here. Those two sentences contradict each other.

> Wherever SimplifyCFG generates a new phi

That's not what I meant: SimplifyCFG could e.g. replace a `switch_enum` with a branch to a payload-block. Having separate argument types for phi/terminator would require SimplifyCFG to replace the payload argument. Whereas right now it doesn't need to. If someone forgets to replace the argument in such a case, it can only be caught by the verifier.

> SILBlockArgument

We don't have a SILBlockArgument class. We have `SILPhiArgument`, which is totally confusing, because it's both a phi argument and a terminator result argument.

Reply:
> > There is never a situation in which a pass cares where a SILArgument comes from that it doesn't need to distinguish phis from terminator results. They are handled in an opposite fashion in all the code I've seen.
> 
> I clearly misunderstood something here. Those two sentences contradict each other.

Forget all the names we currently have. They don't make any sense. Blocks have arguments. I'll call those "block arguments". Sometimes, it doesn't matter where a "block argument" comes from. That's fine. You can just use a BlockArgument API. But often it does matter where the argument comes from. When it does, you need this switch:
```
(blockArgument : BlockArgument)
switch blockArgument {
  case FunctionArgument: ;
  case Phi ;
  case TerminatorResult ;
}
```


This is particularly important for working with value ownership or inserting compensation code. These three block argument types are always handled differently. FunctionArguments usually need to be bypassed. Phis and TerminatorResults have opposite handling because they occur in opposite situations and have completely different meaning. One handles values live into a control flow merge. The other handle values that come into existence after a control flow branch. In situations where we handle phis, we need to know that the argument is not a terminator result and vice-versa.

We need static types for each of these cases because we need to build APIs around them. Those APIs need to be readable and safe. The address lowering pass was impossible to understand until I created the Phi wrapper type. It's still confusing because we don't have a proper TerminatorResult type. We only have terminator result APIs on SILArgument.

A lot of the difficulty in converting passes like SimplifyCFG to OSSA is that the code *implicitly assumes* that it's dealing with one or the other kind of block argument making it impossible to follow the code. Also there are no natural, reasonably safe APIs for working with phis and terminators. It's all hard-coded at a low level.

> I don't understand your comment. In Swift we don't have opcodes, we just have types.

SIL has opcodes. Swift models them as class types (creating redundancy between the vtable and opcode). Ideally, we have three opcodes/class types: FunctionArgument, Phi, TerminatorResult; and one protocol: BlockArgument.

Alternatively, we have one opcode: BlockArgument; and three wrapper struct types.

I don't have a strong preference for either. The former is more efficient, but we seem to be ignoring efficiency given that we're replacing opcodes (integer comparison) with protocol conformance lookup!

I understand you don't want to change the C++ code, and therefore don't want to change SIL opcodes. That's fine, but this document is about how to fix the serious design mistakes in C++, not about what you feel like doing at this time. If no one wants to fix the SIL opcodes, then we can use Swift wrapper types to hide the problem. We're already doing that on the C++ side.

If we don't eliminate the SILFunction and SILPhiArgument SIL opcodes, then we should at least stop bridging them into Swift before the problem gets even worse.
 
> > Wherever SimplifyCFG generates a new phi
> 
> That's not what I meant: SimplifyCFG could e.g. replace a `switch_enum` with a branch to a payload-block. Having separate argument types for phi/terminator would require SimplifyCFG to replace the payload argument. Whereas right now it doesn't need to. If someone forgets to replace the argument in such a case, it can only be caught by the verifier.

That's a good example of why SimplifyCFG is a total mess. It's complete nonsense to replace a `switch_enum` with a `phi`. That can never happen in OSSA. When you delete a `switch_enum` you replace all uses with some replacement value and remove its result. If there needs to be a phi. It had to be there already, completely independent of the switch. Confusing these two opposite concepts is a major reason that SimplifyCFG is such a horror to work with.

Reply:
Currently we have:
```
class Argument {}
final class FunctionArgument : Argument {}
final class BlockArgument : Argument {
  var isPhiArgument: Bool
}
```
and the switch would look like:
```
switch argument {
  case is FunctionArgument: ...
  case let blockArg as BlockArgument:
    if blockArg.isPhiArgument {
      ...
    } else {
      ...
    }
}
```
But we can also make three argument classes:
```
class Argument {}
final class FunctionArgument : Argument {}
class BlockArgument : Argument { }
final class PhiArgument : BlockArgument {}
final class TerminatorResult : BlockArgument {}
```
I don't have a strong opinion on this.
But if we do, I strongly suggest to first fix this in C++ and not introducing wrapper structs (which would just add an unnecessary abstraction layer = unnecessary complexity).

Reply:
These two classes are redundant in practice. 
```
class Argument {}
class BlockArgument : Argument { }
```
I don't think there's any API that belongs on non-function arguments ("BlockArgument") but does not belong on all block arguments. I think it causes bugs to attempt to handle non-function arguments without first determining whether they are phis or terminators.

It's great to fix this in C++ and add class types. But I don't understand why a wrapper struct is more complicated than a class type. The difference in syntax is superficial
`case is Phi` vs `case let phi = argument.asPhi()`